orderbook/src/account_balances.rs

```diff
@@ -178,116 +109,73 @@ impl Web3BalanceFetcher {
     }
 }
 
-fn erc20_balance_query<'a>(
-    batch: &mut CallBatch<&'a Web3Transport>,
+fn erc20_balance_query(
+    batch: &mut CallBatch<Web3Transport>,
     token: ERC20,
     owner: H160,
     spender: H160,
-) -> BoxFuture<'a, SubscriptionValue> {
+) -> impl Future<Output = Result<U256>> {
     let balance = token.balance_of(owner).batch_call(batch);
     let allowance = token.allowance(owner, spender).batch_call(batch);
-
     async move {
-        let (balance, allowance) = futures::join!(balance, allowance);
-
-        let mut value = SubscriptionValue::default();
-        match balance {
-            Ok(balance) => value.balance = Some(balance),
-            Err(_) => tracing::warn!(
-                "Couldn't fetch {:?} balance for {:?}",
-                token.address(),
-                owner
-            ),
-        }
-        match allowance {
-            Ok(allowance) => value.allowance = Some(allowance),
-            Err(_) => tracing::warn!(
-                "Couldn't fetch {:?} allowance from {:?} to {:?}",
-                token.address(),
-                owner,
-                spender
-            ),
-        }
-
-        value
+        let balance = balance.await.context("balance")?;
+        let allowance = allowance.await.context("allowance")?;
+        Ok(balance.min(allowance))
     }
-    .boxed()
 }
 
-fn vault_external_balance_query<'a>(
-    batch: &mut CallBatch<&'a Web3Transport>,
+fn vault_external_balance_query(
+    batch: &mut CallBatch<Web3Transport>,
     vault: BalancerV2Vault,
     token: ERC20,
     owner: H160,
     relayer: H160,
-) -> BoxFuture<'a, SubscriptionValue> {
-    let erc20 = erc20_balance_query(batch, token, owner, vault.address());
+) -> impl Future<Output = Result<U256>> {
+    let balance = token.balance_of(owner).batch_call(batch);
     let approval = vault.has_approved_relayer(owner, relayer).batch_call(batch);
-
     async move {
-        let (value, approval) = futures::join!(erc20, approval);
-        match approval {
-            Ok(true) => value,
-            Ok(false) => SubscriptionValue {
-                allowance: Some(0.into()),
-                ..value
-            },
-            Err(_) => {
-                tracing::warn!(
-                    "Couldn't fetch vault approval from {:?} to {:?}",
-                    owner,
-                    relayer
-                );
-                SubscriptionValue::default()
-            }
-        }
+        Ok(match approval.await.context("allowance")? {
+            true => balance.await.context("balance")?,
```

Comment:
I think this may be slightly wrong. If you have vault relayer approval, then the balance is:
`cmp::min(token.balance_of(owner), token.allowance(owner, vault))`

So I think it should be:
```
let balance = erc20_balance_query(batch, ...);
```

Reply:
Good point. Will fix and see if I can find a good way to add a test for it.